### PR TITLE
(unittest) Added retro_unittest.h

### PR DESCRIPTION
**Pull request by ctult**

### CODE REVIEW

File: libretro-common/include/retro_unittest.h [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+ #define _RUNIT_PRINT_PASS(msg)
+#endif
+static void (*runit_setup)(void) = NULL;
+static void (*runit_teardown)(void) = NULL;
+
+static int
+   runit_run      = 0,
+   runit_assert   = 0,
+   runit_fail     = 0,
+   runit_status   = 0;
```
Comment (heuripedes) [inline libretro-common/include/retro_unittest.h]: Please either use `static int` for every variable or merge them in a single line.

**Comment by heuripedes** [comment]
Can you include a sample file too?

**Comment by Alcaro** [comment]
I agree with the general idea, but this implementation is a mess.

> #define _RUNIT_PRINTF(fmt,args...) fprintf(stderr, fmt, ##args)

That construction [is GCC specific](https://gcc.gnu.org/onlinedocs/cpp/Variadic-Macros.html).

To start with, replace with `#define _RUNIT_PRINTF(...) fprintf(stderr, __VA_ARGS__)`; it's ugly, but at least it's standard. Unfortunately, it requires C99 or C++11; @twinaphex, are C99 dependencies acceptable in optional code?

> static int runit_run = 0;

Large piles of global variables. Static global variables that share name, but not address, between various files. Ew.

> + RUNIT_RUN_SUITE(libretro);

Test code does not belong on the real codepaths. Not even if ifdeffed, it gunks up the code and makes it harder to read.

> +#endif

You forgot the newline. IIRC, that's undefined behaviour according to the standard. (Of course all compilers let it pass, possibly with a warning, but should be fixed anyways.)

This entire scheme feels like a giant overengineered mess to me. I'd prefer scrapping it entirely.

My favourite solution would be making a shell/Python/Make/whatever script that locates files named something_test.c, cooks up a main file that calls them all, and removes the real main().

Removing main() can be done, for example, via #define main not_really_main and a #undef in the test main file. There is a HAVE_MAIN define that removes main(), but it implies there's another main elsewhere, and we need to remove those mains too, in a way that does not put test code on non-test paths.

We also have a couple of tests already. Their needs (except the test cores) should be considered in whatever scheme we make.

```
$ find -ipath '*test*'
./libretro-db/tests.lua
./libretro-db/testlib.c
./libretro-db/rmsgpack_test.c
./tests
./tests/test_reentrancy.c
./cores/libretro-test
./cores/libretro-test/jni
./cores/libretro-test/jni/Application.mk
./cores/libretro-test/jni/Android.mk
./cores/libretro-test/Makefile
./cores/libretro-test/link.T
./cores/libretro-test/libretro-test.c
./cores/libretro-test-gl
./cores/libretro-test-gl/jni
./cores/libretro-test-gl/jni/Application.mk
./cores/libretro-test-gl/jni/Android.mk
./cores/libretro-test-gl/Makefile
./cores/libretro-test-gl/link.T
./cores/libretro-test-gl/libretro-test.c
./audio/test
./audio/test/kaiser_window.m
./audio/test/snr.c
./audio/test/sinc_test.m
./audio/test/Makefile
./audio/test/test-rate-control.sh
./audio/test/modified_bessel.m
./audio/test/main.c
./libretro-common/utils/udp-test.c
./libretro-common/formats/xml/rxml_test.c
./libretro-common/formats/png/rpng_test.c
./libretro-common/net/net_http_test.c
./libretro-common/file/nbio/nbio_test.c
./ui/drivers/qt/wimp-test
./ui/drivers/qt/wimp-test/main.cpp
./ui/drivers/qt/wimp-test/wimp-test.pro
```

**Comment by ctult** [comment]
Replying to @Alcaro's comment:

On hindsight, I have to agree that this is a really terrible implementation. I was trying to make a C-only solution and ended up with something that cannot really be used for unit testing beyond anything more complex than checking a few functions. It's almost a straight port of Minunit minus all of the time-based things Minunit has.

I feel the best first thing to do is to add assertion checks in the code that get disabled when compiling a release build (like assert() but without the crash-on-fail). That in and of itself would make a good amount of the bugs much more prominent.

The testing

**Comment by ctult** [comment]
Replying to @Alcaro's comment:

On hindsight, I agree that this is a really terrible implementation. I was trying to make a C-only solution and ended up with something that cannot really be used for unit testing beyond anything more complex than checking a few functions. It's almost a straight port of Minunit minus all of the time-based things Minunit has.

I feel the best first thing to do is to add runtime assertion checks in the code that get disabled when compiling a release build (like assert() but without the crash-on-fail). That in and of itself would make a good amount of the bugs much more prominent.

I don't know if a shell/Python/make/whatever script would fit in with the rest of Retroarch, but I'll see what I can do.

**Pull request closed without merging**